can/scripts/viewer.py

```diff
@@ -0,0 +1,602 @@
+    if len(parsed_args.filter) > 0:
+        # print('Adding filter/s', parsed_args.filter)
+        for flt in parsed_args.filter:
+            # print(filter)
+            if ':' in flt:
+                _ = flt.split(':')
+                can_id, can_mask = int(_[0], base=16), int(_[1], base=16)
+            elif '~' in flt:
+                can_id, can_mask = flt.split('~')
+                can_id = int(can_id, base=16) | 0x20000000  # CAN_INV_FILTER
+                can_mask = int(can_mask, base=16) & 0x20000000  # socket.CAN_ERR_FLAG
+            else:
+                raise argparse.ArgumentError(None, 'Invalid filter argument')
+            can_filters.append({'can_id': can_id, 'can_mask': can_mask})
+
+    # Dictionary used to convert between Python values and C structs represented as Python strings.
+    # If the value is 'None' then the message does not contain any data package.
+    #
+    # The struct package is used to unpack the received data.
+    # Note the data is assumed to be in little-endian byte order.
+    # < = little-endian, > = big-endian
+    # x = pad byte
+    # c = char
+    # ? = bool
+    # b = int8_t, B = uint8_t
+    # h = int16, H = uint16
+    # l = int32_t, L = uint32_t
+    # q = int64_t, Q = uint64_t
+    # f = float (32-bits), d = double (64-bits)
+    #
+    # An optional conversion from real units to integers can be given as additional arguments.
+    # In order to convert from raw integer value the real units are multiplied with the values and similarly the values
+    # are divided by the value in order to convert from real units to raw integer values.
+    data_structs = {}  # type: Dict[Union[int, Tuple[int, ...]], Union[struct.Struct, Tuple, None]]
+    if len(parsed_args.decode) > 0:
+        if os.path.isfile(parsed_args.decode[0]):
+            with open(parsed_args.decode[0], 'r') as f:
+                structs = f.readlines()
+        else:
+            structs = parsed_args.decode
+
+        for s in structs:
+            tmp = s.rstrip('\n').split(':')
+
+            # The ID is given as a hex value, the format needs no conversion
+            key, fmt = int(tmp[0], base=16), tmp[1]
+
+            # The scaling
+            scaling = []  # type: list
+            for t in tmp[2:]:
+                # First try to convert to int, if that fails, then convert to a float
+                try:
+                    scaling.append(int(t))
+                except ValueError:
+                    scaling.append(float(t))
+
+            if scaling:
+                data_structs[key] = (struct.Struct(fmt),) + tuple(scaling)
+            else:
+                data_structs[key] = struct.Struct(fmt)
+            # print(data_structs[key])
+
+    ignore_canopen = parsed_args.ignore_canopen
+
+    return parsed_args, can_filters, data_structs, ignore_canopen
+
+
+def main():  # pragma: no cover
+    parsed_args, can_filters, data_structs, ignore_canopen = parse_args(sys.argv[1:])
+
+    config = {}
```

Comment:
Put `single_handle=True` here for Kvaser, otherwise we won't be able to see messages sent from the local machine.